delphi/translators/for2py/rectify.py

```diff
@@ -177,60 +413,136 @@ def handle_tag_header(self, root, parElem):
-                ):
-                    self.parseXMLTree(child, curElem)
-                else:
-                    print(f'In handle_tag_header: "{child.tag}" not handled')
+            if child.text or len(child) > 0:
+                cur_elem = ET.SubElement(current, child.tag, child.attrib)
+
+                try:
+                    error_chk = self.header_child_tags.index(child.tag)
+                except ValueError:
+                    assert False, f'In handle_tag_header: "{child.tag}" not handled'
+
+                self.parseXMLTree(child, cur_elem, current, parent, traverse)
             else:
                 if child.tag == "subroutine-stmt":
-                    parElem.attrib.update(child.attrib)
-                elif child.tag == "loop-control" or child.tag == "label":
-                    curElem = ET.SubElement(parElem, child.tag, child.attrib)
+                    current.attrib.update(child.attrib)
+                elif child.tag in self.header_child_tags:
+                    cur_elem = ET.SubElement(current, child.tag, child.attrib)
                 else:
-                    print(
-                        f'In handle_tag_header: Empty elements  "{child.tag}" not handled'
-                    )
+                    assert False, f'In handle_tag_header: Empty elements  "{child.tag}" not handled'
+
 
-    def handle_tag_body(self, root, parElem):
+    def handle_tag_body(self, root, current, parent, grandparent, traverse):
         """
             This function handles cleaning up the XML elementss between the body elementss.
             <body>
                 ...
             </body>
         """
+        new_parent = current
+        current.attrib['parent'] = parent.tag
         for child in root:
             self.clean_attrib(child)
-            if child.text:
-                curElem = ET.SubElement(parElem, child.tag, child.attrib)
-                if (
-                    child.tag == "specification"
-                    or child.tag == "statement"
-                    or child.tag == "loop"
-                    or child.tag == "if"
-                ):
-                    self.parseXMLTree(child, curElem)
+            if len(child) > 0 or child.text:
+                cur_elem = ET.SubElement(current, child.tag, child.attrib)
+                if child.tag in self.body_child_tags:
+                    self.parseXMLTree(child, cur_elem, current, parent, traverse)
+
+                    # Handling conditional <goto-stmt>
+                    if traverse == 1:
+                        if (
+                            parent.tag == "if"
+                            and "goto-stmt" in cur_elem.attrib
+                        ):
+                            assert (
+                                    "lbl" in cur_elem.attrib
+                            ), "Label 'lbl' must be present to store the value in the <if> attrib"
+
+                            rand_code = RD.randrange(RANGE)
```

Comment:
I don't understand this code.  In general, randomly generated code makes me nervous; the lack of comments does not help.  I'd like to discuss this code in person to understand what it's doing.

Reply:
@skdebray This is basically for keeping the track of nested statements with a unique code. And, later verify that the statement is an inner statement of another statement. I will add the comment above the code, but, yes, I will explain to you in person why I believe this is needed. If you believe this code can be replaced with another option, I am willing to replace it. I will be on campus today, so I can come to your office later in the afternoon if you are also planning to be on campus or we can discuss on Monday. Meanwhile, I will address other issues and push the updates. Thank you.